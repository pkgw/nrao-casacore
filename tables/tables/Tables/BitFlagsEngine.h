//# BitFlagsEngine.h: Templated virtual column engine to map bit flags to a Bool
//# Copyright (C) 2009
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id: BitFlagsEngine.h 20539 2009-03-20 08:54:16Z gervandiepen $

#ifndef TABLES_BITFLAGSENGINE_H
#define TABLES_BITFLAGSENGINE_H

//# Includes
#include <tables/Tables/BaseMappedArrayEngine.h>

namespace casa { //# NAMESPACE CASA - BEGIN


  // <summary>
  // Templated virtual column engine to map bit flags to a Bool.
  // </summary>

  // <use visibility=export>

  // <reviewed reviewer="Gareth Hunt" date="94Nov17" tests="">
  // </reviewed>

  // <prerequisite>
  //# Classes you should understand before using this one.
  //   <li> VirtualColumnEngine
  //   <li> VirtualArrayColumn
  // </prerequisite>

  // <synopsis> 
  // BitFlagsEngine is a virtual column engine which scales an array
  // of one type to another type to save disk storage.
  // This resembles the classic AIPS compress method which scales the
  // data from float to short.
  // The scale factor and offset value can be given in two ways:
  // <ul>
  //  <li> As a fixed value which is used for all arrays in the column.
  //  <li> As the name of a column. In this way each array in a
  //         column can have its own scale and offset value.
  //         The scale and offset value in a row must be put before
  //         the array is put and should not be changed anymore.
  // </ul>
  // It is also possible to have a variable scale factor with a fixed offset
  // value.
  // As in FITS the scale and offset values are used as:
  // <br><src> True_value = Stored_value * scale + offset; </src>
  //
  // An engine object should be used for one column only, because the stored
  // column name is part of the engine. If it would be used for more than
  // one column, they would all share the same stored column.
  // When the engine is bound to a column, it is checked if the name
  // of that column matches the given virtual column name.
  //
  // The engine can be used for a column containing any kind of array
  // (thus direct or indirect, fixed or variable shaped)) as long as the
  // virtual array can be stored in the stored array. Thus a fixed shaped
  // virtual can use a variable shaped stored, but not vice versa.
  // A fixed shape indirect virtual can use a stored with direct arrays.
  //
  // This class can also serve as an example of how to implement
  // a virtual column engine.
  // </synopsis> 

  // <motivation>
  // This class allows to store data in a smaller representation.
  // It is needed to resemble the classic AIPS compress option.
  // It adds the scale and offset value on a per row basis.
  //
  // Because the engine can serve only one column, it was possible to
  // combine the engine and the column functionality in one class.
  // This has been achieved using multiple inheritance.
  // The advantage of this is that only one templated class is used,
  // so less template instantiations are needed.
  // </motivation>

  // <example>
  // <srcblock>
  // // Create the table description and 2 columns with indirect arrays in it.
  // // The Int column will be stored, while the double will be
  // // used as virtual.
  // TableDesc tableDesc ("", TableDesc::Scratch);
  // tableDesc.addColumn (ArrayColumnDesc<Int> ("storedArray"));
  // tableDesc.addColumn (ArrayColumnDesc<double> ("virtualArray"));
  //
  // // Create a new table using the table description.
  // SetupNewTable newtab (tableDesc, "tab.data", Table::New);
  //
  // // Create the array scaling engine to scale from double to Int
  // // and bind it to the double column.
  // // Create the table.
  // BitFlagsEngine<Int> scalingEngine("virtualArray",
  //                                             "storedArray", 10);
  // newtab.bindColumn ("virtualArray", scalingEngine);
  // Table table (newtab);
  //
  // // Store a 3-D array (with dim. 2,3,4) into each row of the column.
  // // The shape of each array in the column is implicitly set by the put
  // // function. This will also set the shape of the underlying Int array.
  // ArrayColumn data (table, "virtualArray");
  // Array<double> someArray(IPosition(4,2,3,4));
  // someArray = 0;
  // for (uInt i=0, i<10; i++) {          // table will have 10 rows
  //     table.addRow();
  //     data.put (i, someArray)
  // }
  // </srcblock>
  // </example>

  // <templating arg=VirtualType>
  //  <li> only suited for built-in numerics data types
  // </templating>
  // <templating arg=StoredType>
  //  <li> only suited for built-in numerics data types
  // </templating>

  template<typename StoredType> class BitFlagsEngine : public BaseMappedArrayEngine<Bool, StoredType>
  {
    //# Make members of parent class known.
  public:
    using BaseMappedArrayEngine<Bool,StoredType>::virtualName;
  protected:
    using BaseMappedArrayEngine<Bool,StoredType>::storedName;
    using BaseMappedArrayEngine<Bool,StoredType>::table;
    using BaseMappedArrayEngine<Bool,StoredType>::roColumn;
    using BaseMappedArrayEngine<Bool,StoredType>::rwColumn;
    using BaseMappedArrayEngine<Bool,StoredType>::setNames;

  public:
    // Construct an engine to scale all arrays in a column with
    // the given offset and scale factor.
    // StoredColumnName is the name of the column where the scaled
    // data will be put and must have data type StoredType.
    // The virtual column using this engine must have data type Bool.
    BitFlagsEngine (const String& virtualColumnName,
                    const String& storedColumnName,
                    StoredType readMask=StoredType(0xffffffff),
                    StoredType writeMask=1);

    // Construct from a record specification as created by getmanagerSpec().
    BitFlagsEngine (const Record& spec);

    // Destructor is mandatory.
    ~BitFlagsEngine();

    // Return the type name of the engine (i.e. its class name).
    virtual String dataManagerType() const;

    // Get the name given to the engine (is the virtual column name).
    virtual String dataManagerName() const;
  
    // Record a record containing data manager specifications.
    virtual Record dataManagerSpec() const;

    // Get data manager properties that can be modified.
    // These are READMASK and WRITEMASK.
    // It is a subset of the data manager specification.
    virtual Record getProperties() const;

    // Modify data manager properties.
    // Thety can be READMASK and WRITEMASK.
    virtual void setProperties (const Record& spec);

    // Return the name of the class.
    // This includes the names of the template arguments.
    static String className();

    // Register the class name and the static makeObject "constructor".
    // This will make the engine known to the table system.
    // The automatically invoked registration function in DataManReg.cc
    // contains BitFlagsEngine<Int>.
    // Any other instantiation of this class must be registered "manually"
    // (or added to DataManReg.cc).
    static void registerClass();

  private:
    // Copy constructor is only used by clone().
    // (so it is made private).
    BitFlagsEngine (const BitFlagsEngine<StoredType>&);

    // Assignment is not needed and therefore forbidden
    // (so it is made private and not implemented).
    BitFlagsEngine<StoredType>& operator= (const BitFlagsEngine<StoredType>&);

    // Clone the engine object.
    DataManager* clone() const;

    // Initialize the object for a new table.
    // It defines the keywords containing the engine parameters.
    void create (uInt initialNrrow);

    // Preparing consists of setting the writable switch and
    // adding the initial number of rows in case of create.
    // Furthermore it reads the keywords containing the engine parameters.
    void prepare();

    // Get an array in the given row.
    // This will scale and offset from the underlying array.
    void getArray (uInt rownr, Array<Bool>& array);

    // Put an array in the given row.
    // This will scale and offset to the underlying array.
    void putArray (uInt rownr, const Array<Bool>& array);

    // Get a section of the array in the given row.
    // This will scale and offset from the underlying array.
    void getSlice (uInt rownr, const Slicer& slicer, Array<Bool>& array);

    // Put into a section of the array in the given row.
    // This will scale and offset to the underlying array.
    void putSlice (uInt rownr, const Slicer& slicer,
		   const Array<Bool>& array);

    // Get an entire column.
    // This will scale and offset from the underlying array.
    void getArrayColumn (Array<Bool>& array);

    // Put an entire column.
    // This will scale and offset to the underlying array.
    void putArrayColumn (const Array<Bool>& array);

    // Get some array values in the column.
    // This will scale and offset from the underlying array.
    virtual void getArrayColumnCells (const RefRows& rownrs,
				      Array<Bool>& data);

    // Put some array values in the column.
    // This will scale and offset to the underlying array.
    virtual void putArrayColumnCells (const RefRows& rownrs,
				      const Array<Bool>& data);

    // Get a section of all arrays in the column.
    // This will scale and offset from the underlying array.
    void getColumnSlice (const Slicer& slicer, Array<Bool>& array);

    // Put a section of all arrays in the column.
    // This will scale and offset to the underlying array.
    void putColumnSlice (const Slicer& slicer, const Array<Bool>& array);

    // Get a section of some arrays in the column.
    // This will scale and offset from the underlying array.
    virtual void getColumnSliceCells (const RefRows& rownrs,
				      const Slicer& slicer,
				      Array<Bool>& data);

    // Put into a section of some arrays in the column.
    // This will scale and offset to the underlying array.
    virtual void putColumnSliceCells (const RefRows& rownrs,
				      const Slicer& slicer,
				      const Array<Bool>& data);

    // Map bit flags array to Bool array.
    // This is meant when reading an array from the stored column.
    void mapOnGet (Array<Bool>& array,
                   const Array<StoredType>& stored);

    // Map Bool array to bit flags array.
    // This is meant when writing an array into the stored column.
    void mapOnPut (const Array<Bool>& array,
                   Array<StoredType>& stored);

    // Functor to and an array and mask and convert to Bool.
    struct FlagsToBool : public std::unary_function<StoredType,Bool>
    {
      explicit FlagsToBool(StoredType readMask) : itsMask(readMask) {}
      Bool operator() (StoredType value) const
        { return (value & itsMask) != 0; }
    private:
      StoredType itsMask;
    };
    // Functor to convert Bools to flags using a mask.
    // By default only bit 0 is set.
    // Flag bits not affected are kept.
    struct BoolToFlags : public std::binary_function<Bool,StoredType,StoredType>
    {
      explicit BoolToFlags(StoredType writeMask) : itsMask(writeMask) {}
      StoredType operator() (Bool flag, StoredType value) const
        { return (flag ? value&itsMask : value); }
    private:
      StoredType itsMask;
    };

  public:
    // Define the "constructor" to construct this engine when a
    // table is read back.
    // This "constructor" has to be registered by the user of the engine.
    // If the engine is commonly used, its registration can be added
    // to the registerAllCtor function in DataManReg.cc. 
    // That function gets automatically invoked by the table system.
    static DataManager* makeObject (const String& dataManagerType,
				    const Record& spec);

  private:
    StoredType itsReadMask;
    StoredType itsWriteMask;
  };



} //# NAMESPACE CASA - END

#ifndef CASACORE_NO_AUTO_TEMPLATES
#include <tables/Tables/BitFlagsEngine.tcc>
#endif //# CASACORE_NO_AUTO_TEMPLATES
#endif
