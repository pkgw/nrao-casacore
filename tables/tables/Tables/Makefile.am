include $(top_srcdir)/Makefile.am.common
BUILT_SOURCES = RecordGram.lcc RecordGram.ycc TableGram.lcc TableGram.ycc
