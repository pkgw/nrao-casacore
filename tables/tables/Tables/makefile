# $Id: makefile 17783 2004-08-25 00:30:19Z cvsmgr $

# Use the generic AIPS++ class implementation makefile.
#------------------------------------------------------
include $(word 1, $(AIPSPATH))/code/install/makefile.imp
