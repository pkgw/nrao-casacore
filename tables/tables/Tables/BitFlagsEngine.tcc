//# BitFlagsEngine.tcc: Templated virtual column engine to map bit flags to a Bool
//# Copyright (C) 2009
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id: BitFlagsEngine.tcc 20539 2009-03-20 08:54:16Z gervandiepen $

//# Includes
#include <tables/Tables/BitFlagsEngine.h>
#include <tables/Tables/Table.h>
#include <tables/Tables/ScalarColumn.h>
#include <tables/Tables/ArrayColumn.h>
#include <tables/Tables/ColumnDesc.h>
#include <tables/Tables/TableRecord.h>
#include <tables/Tables/DataManError.h>
#include <casa/Arrays/Array.h>
#include <casa/Arrays/ArrayMath.h>
#include <casa/Containers/Record.h>
#include <casa/BasicSL/String.h>
#include <casa/Utilities/ValTypeId.h>


namespace casa { //# NAMESPACE CASA - BEGIN

  template<typename T>
  BitFlagsEngine<T>::BitFlagsEngine (const String& virtualColumnName,
                                     const String& storedColumnName,
                                     T readMask, T writeMask)
  : BaseMappedArrayEngine<Bool,T> (virtualColumnName, storedColumnName),
    itsReadMask  (readMask),
    itsWriteMask (writeMask)
  {}

  template<typename T>
  BitFlagsEngine<T>::BitFlagsEngine (const Record& spec)
  : BaseMappedArrayEngine<Bool,T> (),
    itsReadMask  (T(0xffffffff)),
    itsWriteMask (1)
  {
    if (spec.isDefined("SOURCENAME")  &&  spec.isDefined("TARGETNAME")) {
      setNames (spec.asString("SOURCENAME"), spec.asString("TARGETNAME"));
      setProperties (spec);
    }
  }

  template<typename T>
  BitFlagsEngine<T>::BitFlagsEngine (const BitFlagsEngine<T>& that)
  : BaseMappedArrayEngine<Bool,T> (that),
    itsReadMask  (that.itsReadMask),
    itsWriteMask (that.itsWriteMask)
  {}

  template<typename T>
  BitFlagsEngine<T>::~BitFlagsEngine()
  {}

  //# Clone the engine object.
  template<typename T>
  DataManager* BitFlagsEngine<T>::clone() const
  {
    DataManager* dmPtr = new BitFlagsEngine<T> (*this);
    return dmPtr;
  }


  //# Return the type name of the engine (i.e. its class name).
  template<typename T>
  String BitFlagsEngine<T>::dataManagerType() const
  {
    return className();
  }
  //# Return the class name.
  //# Get the data type names using class ValType.
  template<typename T>
  String BitFlagsEngine<T>::className()
  {
    return "BitFlagsEngine<" + valDataTypeId (static_cast<T*>(0));
  }

  template<typename T>
  String BitFlagsEngine<T>::dataManagerName() const
  {
    return virtualName();
  }

  template<typename T>
  Record BitFlagsEngine<T>::dataManagerSpec() const
  {
    Record spec = getProperties();
    spec.define ("SOURCENAME", virtualName());
    spec.define ("TARGETNAME", storedName());
    return spec;
  }

  template<typename T>
  Record BitFlagsEngine<T>::getProperties() const
  {
    Record spec;
    spec.define ("READMASK",  Int(itsReadMask));
    spec.define ("WRITEMASK", Int(itsWriteMask));
    return spec;
  }

  template<typename T>
  void BitFlagsEngine<T>::setProperties (const Record& spec)
  {
    if (spec.isDefined("READMASK")) {
      itsReadMask  = T(spec.asInt("READMASK"));
    }
    if (spec.isDefined("WRITEMASK")) {
      itsWriteMask = T(spec.asInt("WRITEMASK"));
    }
  }

  template<typename T>
  DataManager* BitFlagsEngine<T>::makeObject (const String&,
                                              const Record& spec)
  {
    DataManager* dmPtr = new BitFlagsEngine<T>(spec);
    return dmPtr;
  }
  template<typename T>
  void BitFlagsEngine<T>::registerClass()
  {
    DataManager::registerCtor (className(), makeObject);
  }


  template<typename T>
  void BitFlagsEngine<T>::create (uInt initialNrrow)
  {
    BaseMappedArrayEngine<Bool,T>::create (initialNrrow);
    // Store the various parameters as keywords in this column.
    TableColumn thisCol (table(), virtualName());
    thisCol.rwKeywordSet().define ("_BitFlagsEngine_ReadMask",
				   Int(itsReadMask));
    thisCol.rwKeywordSet().define ("_BitFlagsEngine_WriteMask",
				   Int(itsWriteMask));
  }

  template<typename T>
  void BitFlagsEngine<T>::prepare()
  {
    BaseMappedArrayEngine<Bool,T>::prepare();
    ROTableColumn thisCol (table(), virtualName());
    itsReadMask  = T(thisCol.keywordSet().asInt("_BitFlagsEngine_ReadMask"));
    itsWriteMask = T(thisCol.keywordSet().asInt("_BitFlagsEngine_WriteMask"));
  }


  template<typename T>
  void BitFlagsEngine<T>::getArray (uInt rownr, Array<Bool>& array)
  {
    Array<T> target(array.shape());
    roColumn().get (rownr, target);
    mapOnGet (array, target);
  }
  template<typename T>
  void BitFlagsEngine<T>::putArray (uInt rownr, const Array<Bool>& array)
  {
    Array<T> target(array.shape());
    mapOnPut (array, target);
    rwColumn().put (rownr, target);
  }

  template<typename T>
  void BitFlagsEngine<T>::getSlice (uInt rownr, const Slicer& slicer,
                                    Array<Bool>& array)
  {
    Array<T> target(array.shape());
    roColumn().getSlice (rownr, slicer, target);
    mapOnGet (array, target);
  }
  template<typename T>
  void BitFlagsEngine<T>::putSlice (uInt rownr, const Slicer& slicer,
                                    const Array<Bool>& array)
  {
    Array<T> target(array.shape());
    mapOnPut (array, target);
    rwColumn().putSlice (rownr, slicer, target);
  }

  template<typename T>
  void BitFlagsEngine<T>::getArrayColumn (Array<Bool>& array)
  {
    Array<T> target(array.shape());
    roColumn().getColumn (target);
    mapOnGet (array, target);
  }
  template<typename T>
  void BitFlagsEngine<T>::putArrayColumn (const Array<Bool>& array)
  {
    Array<T> target(array.shape());
    mapOnPut (array, target);
    rwColumn().putColumn (target);
  }

  template<typename T>
  void BitFlagsEngine<T>::getArrayColumnCells (const RefRows& rownrs,
                                               Array<Bool>& array)
  {
    Array<T> target(array.shape());
    roColumn().getColumnCells (rownrs, target);
    mapOnGet (array, target);
  }
  template<typename T>
  void BitFlagsEngine<T>::putArrayColumnCells (const RefRows& rownrs,
                                               const Array<Bool>& array)
  {
    Array<T> target(array.shape());
    mapOnPut (array, target);
    rwColumn().putColumnCells (rownrs, target);
  }

  template<typename T>
  void BitFlagsEngine<T>::getColumnSlice (const Slicer& slicer,
                                          Array<Bool>& array)
  {
    Array<T> target(array.shape());
    roColumn().getColumn (slicer, target);
    mapOnGet (array, target);
  }
  template<typename T>
  void BitFlagsEngine<T>::putColumnSlice (const Slicer& slicer,
                                          const Array<Bool>& array)
  {
    Array<T> target(array.shape());
    mapOnPut (array, target);
    rwColumn().putColumn (slicer, target);
  }

  template<typename T>
  void BitFlagsEngine<T>::getColumnSliceCells (const RefRows& rownrs,
                                               const Slicer& slicer,
                                               Array<Bool>& array)
  {
    Array<T> target(array.shape());
    roColumn().getColumnCells (rownrs, slicer, target);
    mapOnGet (array, target);
  }
  template<typename T>
  void BitFlagsEngine<T>::putColumnSliceCells (const RefRows& rownrs,
                                               const Slicer& slicer,
                                               const Array<Bool>& array)
  {
    Array<T> target(array.shape());
    mapOnPut (array, target);
    rwColumn().putColumnCells (rownrs, slicer, target);
  }

  template<typename T>
  void BitFlagsEngine<T>::mapOnGet (Array<Bool>& array,
                                    const Array<T>& stored)
  {
    arrayTransform (stored, array, FlagsToBool(itsReadMask));
  }

  template<typename T>
  void BitFlagsEngine<T>::mapOnPut (const Array<Bool>& array,
                                    Array<T>& stored)
  {
    arrayTransformInPlace (stored, array, BoolToFlags(itsWriteMask));
  }

} //# NAMESPACE CASA - END
