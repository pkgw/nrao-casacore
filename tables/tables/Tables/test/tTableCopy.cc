//# tTableCopy.cc: Test program for copying tables
//# Copyright (C) 2006
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id: tTableCopy.cc 20329 2008-06-06 07:59:22Z gervandiepen $

#include <tables/Tables.h>
#include <stdexcept>
#include <iostream>
using namespace casa;
using namespace std;

// Remove the dirname from the table name in an error message.
String removeDir (const String& msg)
{
  String s = msg;
  s.gsub (Regex("/.*/t"), "t");
  return s;
}

int main (int argc, const char* argv[])
{
  Table::TableType ttyp = Table::Plain;
  if (argc > 1  &&  String(argv[1]) == String("m")) {
    ttyp = Table::Memory;
  }
  Bool noRows = False;
  if (argc > 2  &&  String(argv[2]) == String("n")) {
    noRows = True;
  }
  try {
    TableDesc td("", "1", TableDesc::Scratch);
    td.comment() = "A SDMemTable";
    td.addColumn(ScalarColumnDesc<Int>("Test"));
    // Now create a new table from the description.
    SetupNewTable aNewTab("tTableCopy_tmp.tbl", td, Table::New);
    Table tabl(aNewTab, ttyp, 0);
    tabl.addRow();
    
    TableDesc std("", "1", TableDesc::Scratch);
    std.comment() = "A SubTable";
    SetupNewTable newTab(tabl.tableName()+"/SUBTABLE", std, Table::New);
    Table stabl(newTab, ttyp, 0);
    stabl.addRow();
    tabl.rwKeywordSet().defineTable("SUBTABLE", stabl);
    
    tabl.copy("tTableCopy_tmp.newtbl", Table::New, noRows);
    Table t("tTableCopy_tmp.newtbl");

    cout << removeDir(tabl.tableName()) << endl;
    cout << removeDir(tabl.keywordSet().asTable("SUBTABLE").tableName())
	 << endl;
    cout << removeDir(t.tableName()) << endl;
    cout << removeDir(t.keywordSet().asTable("SUBTABLE").tableName()) << endl;
    cout << tabl.nrow() << ' ' << stabl.nrow() << ' '
	 << t.nrow() << ' ' << t.keywordSet().asTable("SUBTABLE").nrow() << ' '
	 << tabl.tableType() << ' ' << t.tableType()
	 << endl;
  } catch (exception& x) {
    cout << x.what() << endl;
    return 1;
  }
  return 0;
}
