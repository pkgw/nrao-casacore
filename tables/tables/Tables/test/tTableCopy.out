tTableCopy_tmp.tbl
tTableCopy_tmp.tbl/SUBTABLE
tTableCopy_tmp.newtbl
tTableCopy_tmp.newtbl/SUBTABLE
1 1 1 1 0 0
tTableCopy_tmp.tbl
tTableCopy_tmp.tbl/SUBTABLE
tTableCopy_tmp.newtbl
tTableCopy_tmp.newtbl/SUBTABLE
1 1 1 1 1 0
tTableCopy_tmp.tbl
tTableCopy_tmp.tbl/SUBTABLE
tTableCopy_tmp.newtbl
tTableCopy_tmp.newtbl/SUBTABLE
1 1 0 0 0 0
tTableCopy_tmp.tbl
tTableCopy_tmp.tbl/SUBTABLE
tTableCopy_tmp.newtbl
tTableCopy_tmp.newtbl/SUBTABLE
1 1 0 0 1 0
